### PR TITLE
feat: add oracle miss counters and ibc inflows grpc api

### PR DESCRIPTION
**Pull request by gsk967**
## Description

closes: #XXXX

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

## Summary by CodeRabbit

- **New Features**
  - Introduced a new query method for tracking oracle vote miss counts by validators.
  - Expanded gRPC service capabilities to include queries for IBC inflow information and quota expiration times.

- **Bug Fixes**
  - Improved the `MissCounters` function to handle specific validator queries and return accurate miss counts.

- **Documentation**
  - Updated service definitions and comments to reflect new query functionalities.

- **Refactor**
  - Enhanced existing query functions to support new inflow and outflow data retrieval requirements.

### CODE REVIEW

File: x/oracle/keeper/grpc_query.go [hunk 1]
```diff
@@ -362,3 +362,38 @@ func (q querier) ExgRatesWithTimestamp(
...
 	return &types.QueryExgRatesWithTimestampResponse{ExgRates: exgRates}, nil
 }
+
+// MissCounters implements types.QueryServer.
+func (q querier) MissCounters(goCtx context.Context, req *types.QueryMissCounters) (*types.QueryMissCountersResponse,
+	error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
```
Comment (robert-zaremba) [inline x/oracle/keeper/grpc_query.go:372]:
```suggestion
```

File: x/oracle/keeper/grpc_query.go [hunk 2]
```diff
@@ -362,3 +362,38 @@ func (q querier) ExgRatesWithTimestamp(
...
+			MissCounter: missCounter,
+		})
+	} else {
+		q.IterateMissCounters(ctx, func(val sdk.ValAddress, u uint64) bool {
+			pfMissCounts = append(pfMissCounts, types.PriceMissCounter{
+				Validator:   val.String(),
+				MissCounter: u,
+			})
+			return false
+		})
```
Comment (robert-zaremba) [inline x/oracle/keeper/grpc_query.go:393]: we have a helper function to load all values: `store.LoadAll`

File: x/uibc/quota/grpc_query.go [hunk 3]
```diff
@@ -57,3 +57,38 @@ func (q Querier) AllOutflows(goCtx context.Context, _ *uibc.QueryAllOutflows) (
...
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	var amount sdk.Dec
+	if len(req.Denom) != 0 {
+		tokenInflow := q.Keeper(&ctx).GetTokenInflow(req.Denom)
+		amount = tokenInflow.Amount
+	} else {
+		amount = q.Keeper(&ctx).GetInflowSum()
+	}
+	return &uibc.QueryInflowsResponse{Amount: amount}, nil
+}
```
Comment (robert-zaremba) [inline x/uibc/quota/grpc_query.go:94]: maybe we should have one query, that will return both total and per token:
```
type QueryInflowsResponse struct {
  Sum sdk.Dec
  PerToken []sdk.DecCoin  // will be only one element if token argument is specified in the query
}
```
What do you think?

File: x/uibc/quota/grpc_query.go [hunk 4]
```diff
@@ -57,3 +57,34 @@ func (q Querier) AllOutflows(goCtx context.Context, _ *uibc.QueryAllOutflows) (
...
+	} else {
+		inflows, err = q.Keeper(&ctx).GetAllInflows()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &uibc.QueryInflowsResponse{Sum: q.Keeper(&ctx).GetInflowSum(), Inflows: inflows}, nil
+}
+
+// QuotaExpires returns the current ibc quota end time.
```
Comment (robert-zaremba) [inline x/uibc/quota/grpc_query.go:80]:
```suggestion
// QuotaExpires returns the current ibc quota expire time.
```

File: x/oracle/keeper/grpc_query.go (hunk 2 above)
Comment (gsk967) [reply x/oracle/keeper/grpc_query.go:389]: `store.LoadAll` will not useful here, because we are storing valAddr as key in this case

**Review by robert-zaremba — APPROVED** [review]

**Pull request merged**